modules/sr/robot3/camera.py

```diff
@@ -1,229 +1,190 @@
-        self.camera.recognitionEnable(self._timestep)
-
-        self._lock = lock
+    def firmware_version(self) -> str | None:
+        """Get the firmware version of the board."""
+        return '1.0'
 
     def see(self, *, eager: bool = True) -> list[Marker]:
         """
-        Identify items which the camera can see and return a list of `Marker`
-        instances describing them.
-        """
-        # Webots appears not to like it if you try to hang on to a
-        # `CameraRecognitionObject` after another time-step has passed. However
-        # because we advance the time-steps in a background thread we're likely
-        # to do that all the time. In order to counter that we have our `Robot`
-        # pass down its time-step lock so that we can hold that while we do the
-        # processing. The objects which we pass back to the caller are safe to
-        # use because they don't refer to Webots' objects at all.
-        with self._lock:
-            self._webot.step(self._timestep)
-            return self._see()
+        Capture an image and identify fiducial markers.
 
-    def _see(self) -> list[Marker]:
-        object_infos = {}
-        recognition_objects = []
-
-        for recognition_object in self.camera.getRecognitionObjects():
-            recognition_model = recognition_object.getModel()
-            if isinstance(recognition_model, bytes):
-                recognition_model = recognition_model.decode(errors='replace')
-            marker_info = parse_marker_info(
-                recognition_model,
-            )
-            if marker_info:
-                recognition_objects.append(recognition_object)
-                object_infos[recognition_object.getId()] = marker_info
-
-        tokens = tokens_from_objects(
-            recognition_objects,
-            lambda o: object_infos[o.getId()].size_m,
-            lambda o: object_infos[o.getId()].get_token_class(),
-        )
-
-        when = self._webot.getTime()
-
-        markers = []
-
-        for token, recognition_object in tokens:
-            marker_info = object_infos[recognition_object.getId()]
-            for face in token.visible_faces():
-                markers.append(Marker(face, marker_info, when))
-
-        return markers
+        :param eager: Process the pose estimations of markers immediately,
+            currently unused.
+        :returns: list of markers that the camera could see.
+        """
+        raw_markers: list[Marker] = []
+        recognition_objects: dict[str, dict[str, Recognition]] = defaultdict(dict)
+
+        with self._capture():
+            for recognition_object in self._camera.getRecognitionObjects():
+                # Get the object's assigned "model" value, the marker has 5 detection points:
+                # the marker itself and the 4 corners, named in the form <id>_<location>
+                tag_uid, tag_name, tag_point = str(recognition_object.getModel()).split('_')
```

Comment:
Note that because `getModel` returns `bytes` on R2022b, calling `str` on that gives a string like `"b'1234'"`. We should probably extract a function `force_text` which does the conversion without adding the bytes `repr` to the string.

Reply:
8181cdb fixes this, making marker recognition work on R2022b.